wallet/src/Pos/Wallet/Web/Account.hs

```diff
@@ -46,24 +49,46 @@ type AccountMode ctx m =
 myRootAddresses :: MonadKeysRead m => m [CId Wal]
 myRootAddresses = encToCId <<$>> getSecretKeysPlain
 
+getKeyById
+    :: AccountMode ctx m
+    => CId Wal
+    -> m (Either PublicKey EncryptedSecretKey)
+getKeyById walletId = do
+    secretKeys <- getSecretKeys
+    case getSKByIdPure secretKeys walletId of
+        Just secretKey -> return $ Right secretKey
+        Nothing -> do
+            -- There's no secret key for 'walletId', this wallet is an external one.
+            publicKeys <- getPublicKeys
+            case getPKByIdPure publicKeys walletId of
+                Just publicKey -> return $ Left publicKey
+                Nothing -> throwM . InternalError $
+                     sformat ("getKeyById: there's no public key for wallet "%build) walletId
+
+getPKByIdPure
+    :: AllUserPublics
+    -> CId Wal
+    -> Maybe PublicKey
+getPKByIdPure (AllUserPublics publicKeys) walletId =
+    -- If 'walletId' corresponds to classic (internal) wallet, there's no public key for it.
+    find (\pk -> (encodeCType . makePubKeyAddressBoot $ pk) == walletId) publicKeys
```

Comment:
I generally find a mix of operators and parentheses to be a little confusing; you can simplify this a bit with:

```haskell
find ((walletId ==) . encodeCType . makePubKeyAddressRoot) publicKeys
```

if you want to go point-free, or

```haskell
find (\pk -> walletId == encodeCType (makePubKeyAddressBoot pk)) publicKeys
```

if you like points :)